hail/src/main/scala/is/hail/expr/ir/functions/StringFunctions.scala

```diff
@@ -133,50 +135,82 @@ object StringFunctions extends RegistryFunctions {
     registerIR("sliceLeft", TString, TInt32, TString) { (s, end) => invoke("slice", TString, s, I32(0), end) }
 
-    registerCode("str", tv("T"), TString, null) { case (r, rt, (aT, a)) =>
+    registerCode("str", tv("T"), TString, (pt: PType) => PCanonicalString(pt.required)) { case (r, rt, (aT, a)) =>
       val annotation = boxArg(r, aT)(a)
       val str = r.mb.getType(aT.virtualType).invoke[Any, String]("str", annotation)
       unwrapReturn(r, rt)(str)
     }
 
-    registerCodeWithMissingness("showStr", tv("T"), TInt32, TString, null) { case (r, rt, (aT, a), (_, trunc)) =>
+    registerCodeWithMissingness("showStr", tv("T"), TInt32, TString, {
+      (_: PType, truncType: PType) => PCanonicalString(truncType.required)
+    }) { case (r, rt, (aT, a), (_, trunc)) =>
       val annotation = Code(a.setup, a.m).muxAny(Code._null(boxedTypeInfo(aT)), boxArg(r, aT)(a.v))
       val str = r.mb.getType(aT.virtualType).invoke[Any, Int, String]("showStr", annotation, trunc.value[Int])
       EmitCode(trunc.setup, trunc.m, PCode(rt, unwrapReturn(r, rt)(str)))
     }
 
-    registerCodeWithMissingness("json", tv("T"), TString, null) { case (r, rt, (aT, a)) =>
+    registerCodeWithMissingness("json", tv("T"), TString, (_: PType) => PCanonicalString(true)) { case (r, rt, (aT, a)) =>
       val bti = boxedTypeInfo(aT)
       val annotation = Code(a.setup, a.m).muxAny(Code._null(bti), boxArg(r, aT)(a.v))
       val json = r.mb.getType(aT.virtualType).invoke[Any, JValue]("toJSON", annotation)
       val str = Code.invokeScalaObject[JValue, String](JsonMethods.getClass, "compact", json)
       EmitCode(Code._empty, false, PCode(rt, unwrapReturn(r, rt)(str)))
     }
 
-    registerWrappedScalaFunction("reverse", TString, TString, null)(thisClass,"reverse")
-    registerWrappedScalaFunction("upper", TString, TString, null)(thisClass,"upper")
-    registerWrappedScalaFunction("lower", TString, TString, null)(thisClass,"lower")
-    registerWrappedScalaFunction("strip", TString, TString, null)(thisClass,"strip")
-    registerWrappedScalaFunction("contains", TString, TString, TBoolean, null)(thisClass, "contains")
-    registerWrappedScalaFunction("translate", TString, TDict(TString, TString), TString, null)(thisClass, "translate")
-    registerWrappedScalaFunction("startswith", TString, TString, TBoolean, null)(thisClass, "startswith")
-    registerWrappedScalaFunction("endswith", TString, TString, TBoolean, null)(thisClass, "endswith")
-
-    registerWrappedScalaFunction("regexMatch", TString, TString, TBoolean, null)(thisClass, "regexMatch")
-
-    registerWrappedScalaFunction("concat", TString, TString, TString, null)(thisClass, "concat")
-
-    registerWrappedScalaFunction("split", TString, TString, TArray(TString), null)(thisClass, "split")
+    registerWrappedScalaFunction("reverse", TString, TString, (pt: PType) => pt)(thisClass,"reverse")
```

Comment:
these don't seem right -- can't return some weird string ptype here if that's the input. Should be `PCanonicalString()`

Reply:
Why should the transformation change the type of the string? I would expect that a view on a String returns a String, any permutation of that String doesn't change its type, and operations that explode a string either return the atomic unit of that String (Char), or the same String type. For instance, we follow this logic in the Scala functions being invoked:

def upper(s: String): String = s.toUpperCase

So I thought that in the future, when we introduced non-canonical strings, these transformations wouldn't force canonicalization, and we would codegen methods that would respect the input type.

Reply:
What if our string type is something like a `PSlicedString(old string type, start, end)`? We can't construct that thing from the result here. Note that registerWrappedScalaFunction calls `unwrapReturn`, which needs a constructable PType to unwrap.

Reply:
By constructible you mean the serialized representation right (because certainly PSlicedString can be constructed in-memory, else we can't have a PSlicedString).

In these cases I would want the type to remain unchanged. The serialized representation should be governed by the PSlicedString (say .constructible), and we should be able to convert back to PSlicedString after deserialization. This would involve storing the class name alongside the serialized value, such that we can get back to the in-memory representation with a .fromConstructible method. This seems easier to me than handling all of the type conversions to constructible things ourselves.

Reply:
This also allows us to cleanly separate things that are canonical because we just don't have any other information that allows us to make more efficient decisions, from those that are canonical because of some implementation detail of a physical type (such as whether it can be serialized or not, or what its specific in-memory representation is)